### PR TITLE
Fix for hotplug issues

### PR DESCRIPTION
**Pull request by ixhamza**
### Motivation and Context
Fix for hotplug issues

### Description
ZED relies on udev to match vdev guids when the device is removed. However,
udev does not contain the correct blkid information for the vdev due to
which the vdev failed to match when detached and we have to rely on
fault handler to make the device unavailable. This PR allows the vdev to trigger
a Disk Change event whenever a new vdev is added to sync blkid
information with udev.
This PR also changes the device state to REMOVED whenever the device is
unplugged instead of UNAVAIL.

### How Has This Been Tested?
By hotplugging vdevs through QEMU

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ixhamza** [comment]
cc: @amotin, @freqlabs

File: cmd/zed/agents/zfs_retire.c [hunk 1]
```diff
@@ -351,13 +352,19 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
...
-		    replace_with_spare(hdl, zhp, vdev) == B_FALSE) {
+		/* Remove the vdev since device is unplugged */
+		if ((l2arc = ((nvlist_lookup_string(nvl,
+		    FM_EREPORT_PAYLOAD_ZFS_VDEV_TYPE, &devtype) == 0 &&
+		    strcmp(devtype, VDEV_TYPE_L2CACHE) == 0))) ||
+		    (strcmp(class, "resource.fs.zfs.removed") == 0)) {
+			fmd_hdl_debug(hdl, "zpool_vdev_remove_wanted '%s'"
+			    ", ret:%d", devname,
+			    zpool_vdev_remove_wanted(zhp, devname));
+		}
```
Comment (amotin) [inline cmd/zed/agents/zfs_retire.c:363]: I think this block would be more readable if you move `l2arc = ...` out of the if() and zpool_vdev_remove_wanted() out of fmd_hdl_debug().

File: module/os/linux/spl/spl-generic.c [hunk 2]
```diff
@@ -517,6 +519,27 @@ ddi_copyin(const void *from, void *to, size_t len, int flags)
...
+#if defined(_KERNEL) && LINUX_VERSION_CODE >= KERNEL_VERSION(5, 12, 0)
+	struct kobject *disk_kobj = bdev_kobj(bdev);
+#else
+	struct kobject *disk_kobj = &part_to_dev(bdev->bd_part)->kobj;
+#endif
+	int ret = kobject_uevent(disk_kobj, KOBJ_CHANGE);
+	if (ret)
+		pr_warn("ZFS: Sending event '%d' to kobject: '%s' (%p): failed"
+		    "(ret:%d)\n", KOBJ_CHANGE, kobject_name(disk_kobj),
+		    disk_kobj, ret);
```
Comment (amotin) [inline module/os/linux/spl/spl-generic.c:539]: For multi-line `if ()` please use `{}`.

File: module/os/linux/zfs/vdev_disk.c [hunk 3]
```diff
@@ -290,6 +302,14 @@ vdev_disk_open(vdev_t *v, uint64_t *psize, uint64_t *max_psize,
 		bdev = blkdev_get_by_path(v->vdev_path, mode | FMODE_EXCL,
 		    zfs_vdev_holder);
 		if (unlikely(PTR_ERR(bdev) == -ENOENT)) {
+			/*
+			 * There is no point of waiting since device is removed
+			 * explicitly
+			 */
+			if (v->vdev_removed) {
+				break;
+			}
```
Comment (amotin) [inline module/os/linux/zfs/vdev_disk.c:311]: For single line `if ()` body you don't need `{}`.

File: module/zfs/vdev.c [hunk 4]
```diff
@@ -3139,6 +3146,25 @@ vdev_dtl_reassess(vdev_t *vd, uint64_t txg, uint64_t scrub_txg,
 	mutex_exit(&vd->vdev_dtl_lock);
 }
 
+/*
+ * Iterate over all the vdevs except l2arc and spare and process kobj event
+ * depending on the request flag.
+ */
+void vdev_process_kobj_evt(vdev_t *vd, int request) {
```
Comment (amotin) [inline module/zfs/vdev.c:3153]: Please format properly. It complicates function search at lot.

Also, I first started thinking whether you really need two defines for the request, but then ended up thinking why not just create two separate functions to post and to clear? They would look pretty simple and compact, and without extra branching and argument.

File: module/zfs/spa.c [hunk 5]
```diff
@@ -8121,8 +8132,12 @@ spa_async_remove(spa_t *spa, vdev_t *vd)
...
 		vdev_state_dirty(vd->vdev_top);
 
-		/* Tell userspace that the vdev is gone. */
-		zfs_post_remove(spa, vd);
+		if (vd->vdev_zed_event) {
+			vd->vdev_zed_event = B_FALSE;
+		} else {
+			/* Tell userspace that the vdev is gone. */
+			zfs_post_remove(spa, vd);
+		}
```
Comment (amotin) [inline module/zfs/spa.c:8140]: Does thins cause some event loop otherwise, or what?  I am thinking what if some other process rather than one that sent the removal request want to receive this notification?

**Comment by freqlabs** [comment]
For more context:

The first part of this change deals with the fact that udev caches blkid information, which includes fields parsed from vdev labels:
```
#  blkid --probe /dev/sde2
/dev/sde2: VERSION="5000" LABEL="storage" UUID="5081467553143690352" UUID_SUB="8098199377452528262" BLOCK_SIZE="4096" TYPE="zfs_member" USAGE="filesystem" PART_ENTRY_SCHEME="gpt" PART_ENTRY_UUID="550c557e-c90e-11e9-bd70-ac1f6b0adb04" PART_ENTRY_TYPE="516e7cba-6ecf-11d6-8ff8-00022d09712b" PART_ENTRY_NUMBER="2" PART_ENTRY_OFFSET="4194432" PART_ENTRY_SIZE="1949330696" PART_ENTRY_DISK="8:64"
# udevadm info /dev/sde2 | grep ID_FS_
E: ID_FS_VERSION=5000
E: ID_FS_LABEL=storage
E: ID_FS_LABEL_ENC=storage
E: ID_FS_UUID=5081467553143690352
E: ID_FS_UUID_ENC=5081467553143690352
E: ID_FS_UUID_SUB=8098199377452528262
E: ID_FS_UUID_SUB_ENC=8098199377452528262
E: ID_FS_TYPE=zfs_member
E: ID_FS_USAGE=filesystem
```
The `LABEL` field is the pool name, `UUID` is the pool guid, `UUID_SUB` the vdev guid, among other fields.
A change event on the block device is needed for udev to invalidate its cache and re-probe the device whenever we update these in a vdev label, for example when we create/reguid/destroy a pool or add/replace/remove vdevs. Otherwise udev will report incorrect guids to zed and zed will ignore events for the device.

The second part of this change deals with device removal. There is `VDEV_STATE_REMOVED` for a device that has been detached from the system. Both FreeBSD and Linux use this state when an I/O failure indicates the device is gone. On FreeBSD there is also a callback in the kernel when a device is removed, which we use to update the state accordingly without having to wait for I/O to fail. There is no such mechanism on Linux, so we have to rely on udev/zed.
There is currently no way to request `VDEV_STATE_REMOVED` from userland, so it has been added to libzfs using the existing `ZFS_IOC_VDEV_SET_STATE` ioctl, and zed now uses this to mark the vdev as removed when it receives the removal event from udev.

File: module/zfs/spa.c (hunk 5 above)
Comment (ixhamza) [reply module/zfs/spa.c:8140]: Exactly, this is used to prevent the event loop if this is sent from zed. If ```vd->vdev_zed_event``` is set it means that vdev is removed already and zed knows about that.

Comment (amotin) [reply module/zfs/spa.c:8140]: For the reason I specified I'd prefer the loop to be broken on side of the process requesting removal.  But that is just a feeling, I'll leave it to somebody with better understanding of the whole picture.

Comment (ixhamza) [reply module/zfs/spa.c:8140]: I think you are right. I will deal with this from the ZED context.

File: cmd/zed/agents/zfs_retire.c [hunk 6]
```diff
@@ -323,6 +323,9 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
 	boolean_t is_disk;
 	vdev_aux_t aux;
 	uint64_t state = 0;
+	int l2arc = 0;
```
Comment (amotin) [inline cmd/zed/agents/zfs_retire.c:326]: This does not need initialization.

File: cmd/zed/agents/zfs_retire.c [hunk 7]
```diff
@@ -351,13 +354,33 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
...
+
+		/*
+		 * If state removed is requested for already removed vdev,
+		 * its a loopback event from spa_async_remove(). Just
+		 * ignore it.
+		 */
+		if ((vs->vs_state == VDEV_STATE_REMOVED) &&
+		    (state == VDEV_STATE_REMOVED)) {
+			return;
+		}
```
Comment (amotin) [inline cmd/zed/agents/zfs_retire.c:368]: This does not need neither `()` nor `{}`.

File: module/zfs/vdev.c [hunk 8]
```diff
@@ -3139,6 +3146,38 @@ vdev_dtl_reassess(vdev_t *vd, uint64_t txg, uint64_t scrub_txg,
...
+{
+	if (vd->vdev_ops->vdev_op_kobj_evt_post &&
+	    vd->vdev_kobj_flag == B_FALSE) {
+		vd->vdev_kobj_flag = B_TRUE;
+		vd->vdev_ops->vdev_op_kobj_evt_post(vd);
+	}
+
+	for (int c = 0; c < vd->vdev_children; c++) {
+		vdev_post_kobj_evt(vd->vdev_child[c]);
+	}
```
Comment (amotin) [inline module/zfs/vdev.c:3164]: This and below do not need `{}`.

**Review by amotin — APPROVED** [review]
I can continue long about cosmetics, but I have no real objections.

**Comment by tonyhutter** [comment]
Couple of thoughts before I take a look at the code:

We've debated in the past whether or not to remove a vdev from a pool on a udev remove event.  On the surface it seems perfectly logical.  udev tells you the disk has been removed, so remove it from the pool.  Makes sense.  

However, at very large scale (1000s of disks), you often see odd disk behavior that you have to account for.  Like, multiple NVMe drives, on multiple nodes, all simultaneously disappearing and re-appearing within a few seconds of each other.  I've seen SAS disks disappear/re-appear as well.  And as long as the disks come back within the IO timeout amount of time, and are healthy after they come back, then the user wouldn't notice anything (maybe even more so if you're running Lustre on top of ZFS, with it's own timeouts/retries on top of it?). Just to quantify this strangeness a little - I'm seeing over 1000 zed "EC_dev_remove" events across all our disks in the last month (of which only a handful are actual disk replacements).  Had this "remove vdev on udev remove" been enabled, we would have seen tons of pools fault at the same time, causing mass downtime and admin intervention.  Since we didn't have it enabled, there were no issues - all the IOs rode though the intermittent period of disks disappearing/re-appearing.

In short, I think you should make this configurable as a new `autoremove` pool property that is off by default.  We already have an `autoreplace` property, so it makes sense to have an `autoremove` property too.

Also, I'm not in favor of the UNAVAIL -> REMOVED rename.  "UNAVAIL" already has a lot of historical weight (just google "ZFS unavail"), and it might break scripts if we rename it.  UNAVAIL also implies that the drive could come back to life later, which would certainly be accurate in the case of our disappearing/re-appearing drives.

**Comment by amotin** [comment]
@tonyhutter From that perspective I think it would be reasonable for ZED to just not make too quick extra movements like kicking in spares for few minutes after disk disappeared just in case it return.  Any way rebuild will likely take hours, so few minutes do not matter much.  Pardon me if that logic is already there.  But if disk really disappeared according to the block level, it won't handle any new I/O, so hiding it from ZFS is not very productive.  On opposite, we saw that disk references held by ZFS cause actual problems for NVMe and surrounding subsystems, that can't free the resources.

We also have a huge fleet of ZFS systems, having may be not multiple thousands, but up to a thousand disks each, just based on FreeBSD.  And on FreeBSD disk disappearance is immediately reported to ZFS straight within kernel without any additional daemon participation.  Daemon handles only replacement.  And we feel good enough about that.

**Comment by freqlabs** [comment]
We found more issues with removal detection to be fixed.

**Comment by tonyhutter** [comment]
@amotin 

> From that perspective I think it would be reasonable for ZED to just not make too quick extra movements like kicking in spares for few minutes after disk disappeared just in case it return. 

Yea, something like an "autoremove after N seconds" timer would be good.  We would probably want the autoreplace timer to be longer than the IO timeout (plus the IO retry).  The autoremove timer could be added in a future PR.

**Comment by freqlabs** [comment]
Physically removed vdevs are already supposed to transition to VDEV_STATE_REMOVED on IO errors even on Linux, but it doesn't work for several reasons.

`zfs_check_media_change()` for kernels 5.10+ has a bug: it always returns 0.

Beyond that, the `bdev_check_media_change()` KPI it uses doesn't work to check for a hot-plugged device. For NVMe, the `check_events` member of `struct block_device_operations` isn't implemented, so it can't work at all. For the drivers that do implement it, it can only work if the device identifies as removable (/sys/block/*/removable). For example, in a few systems I have handy, a USB flash drive is removable, but a SATA disk in a hot-swap enclosure attached to a SAS controller is not.

This would affect older kernels with `check_media_change()` as well. So in most cases the check to see if a device was removed just isn't working as intended.

**Comment by tonyhutter** [comment]
I tested this a little using a draid1:4d:9c:1s pool of 9 NVMe drives.  I tested powering off/on a NVMe drive using `/sys/bus/pci/slots/<slot>/power` and all seemed to work as expected.  I plan on doing some more real world testing next week with spinning disks in a SAS enclosure.

**Comment by tonyhutter** [comment]
> The issue with doing a set all-vdevs rather than some kind of inheritance, is that the value won't persist through a zpool replace or apply to any newer vdevs.

Hmm... that really diminishes the benefit of making the thresholds per-vdev then.  You can't just "set it and forget it" on the vdevs, as would be the natural user expectation.  To make the thresholds persistent (the common use case), you would have to set them on top-level-vdevs.  If you added a new top-level-vdev, you would have to set thresholds on the new top-level-vdev as well.  And the user would have to know that they could set values on the top-level-vdevs, which is non-standard.

With all that in mind, does it make sense to just have the thresholds be per-pool properities?

**Comment by tonyhutter** [comment]
I just did another test today using multipathed spinning disks in a SAS enclosure.  I first created a three disk mirror pool, then I removed a disk while the pool was idle.  When I removed the disk the pool status did not change... which is the expected behavior considering these are multipath'd disk where the DM device doesn't necessarily go away or post a remove event when you pull out a disk.  So that's good.  I then wrote to the pool, and the pool correctly reported the disk as UNAVAIL and reported IO errors.

**Comment by tonyhutter** [comment]
Next to test:
1. Remove log device
2. Remove cache device
3. Remove spare
4. Remove file-backed vdev
5. Remove faulted vdev

**Comment by tonyhutter** [comment]
Further testing on a VM shows:

1. regular vdevs (faulted or not) correctly get marked as REMOVED on removal.
2. log devices (faulted or not) correctly get marked as REMOVED on removal.
3. spares status doesn't change on removal.
4. cache device status doesn't change on removal
5. File-backed vdevs status doesn't change on removal

I don't know if 5 can ever be fixed, since you won't see a udev event for their deletion.  3-4 should be made to work though.

**Comment by ixhamza** [comment]
@tonyhutter thank you so much for testing these patches. Since spare vdevs do not have a label, I think it would require some extra work to support hotplugging for spares, maybe in future PR. I tested hotplug for cache vdev on qemu with emulated nvme device though and it was working during my testing but maybe I missed something. Can you please share details of your test for cache vdevs?

**Comment by tonyhutter** [comment]
> Can you please share details of your test for cache vdevs?

I created two virtual disks in virt-manager, made a pool, and removed the one that was the cache:

```
$ sudo ./zpool status
  pool: tank
 state: ONLINE
config:

	NAME        STATE     READ WRITE CKSUM
	tank        ONLINE       0     0     0
	  vde       ONLINE       0     0     0
	cache
	  vdf       ONLINE       0     0     0

errors: No known data errors

# Verify cache device exist before removal
$ lsblk /dev/vdf
NAME   MAJ:MIN RM  SIZE RO TYPE MOUNTPOINTS
vdf    252:80   0    1G  0 disk 
├─vdf1 252:81   0 1014M  0 part 
└─vdf9 252:89   0    8M  0 part 

# Remove cache device in virt-manager

# cache device has been removed
[hutter@fedora36 NAS-115238]$ lsblk /dev/vdf
lsblk: /dev/vdf: not a block device

$ sudo ./zpool status
  pool: tank
 state: ONLINE
config:

	NAME        STATE     READ WRITE CKSUM
	tank        ONLINE       0     0     0
	  vde       ONLINE       0     0     0
	cache
	  vdf       ONLINE       0     0     0

errors: No known data errors
```

**Comment by ixhamza** [comment]
@tonyhutter Thank you for your description. I debugged the issue and it seems to be related to virtio disks. Vdev should have devid "/dev/disk/by-id", so it can be identified during removal. However, by default virtio disks do not seem to have devid unless the serial attribute is hardcoded for the specified disk (https://github.com/systemd/systemd/issues/17670). Even if we provide the serial attribute for virtio disks, ```zed``` won't be able to find devid since there is no handling for virtio disks in ```zfs_device_get_devid()```. If you test this on real hardware or emulated disks, it should not be a problem. Normal vdevs won't be affected by this since they are identified by ```pool_guid``` and ```vdev_guid```. Maybe I can dig more into it in future.
@freqlabs

**Comment by tonyhutter** [comment]
@ixhamza  - I went back to testing on real-world NVMe devices and have some new results.  Good news is that the cache device now correctly shows REMOVED on removal:

```
# zpool status
  pool: tank
 state: ONLINE
config:

	NAME         STATE     READ WRITE CKSUM
	tank         ONLINE       0     0     0
	  mirror-0   ONLINE       0     0     0
	    nvme2n1  ONLINE       0     0     0
	    nvme3n1  ONLINE       0     0     0
	logs	
	  nvme5n1    ONLINE       0     0     0
	cache
	  nvme0n1    ONLINE       0     0     0
	spares
	  nvme1n1    AVAIL   

# echo 0 > /sys/bus/pci/slots/0/power 
# zpool status
  pool: tank
 state: ONLINE
status: One or more devices has been removed by the administrator.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Online the device using zpool online' or replace the device with
	'zpool replace'.
config:

	NAME         STATE     READ WRITE CKSUM
	tank         ONLINE       0     0     0
	  mirror-0   ONLINE       0     0     0
	    nvme2n1  ONLINE       0     0     0
	    nvme3n1  ONLINE       0     0     0
	logs	
	  nvme5n1    ONLINE       0     0     0
	cache
	  nvme0n1    REMOVED      0     0     0
	spares
	  nvme1n1    AVAIL   
```
The bad news is that the spares do not show REMOVED:


```
# echo 0 > /sys/bus/pci/slots/1/power 
# zpool status
  pool: tank
 state: ONLINE
status: One or more devices has been removed by the administrator.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Online the device using zpool online' or replace the device with
	'zpool replace'.
config:

	NAME         STATE     READ WRITE CKSUM
	tank         ONLINE       0     0     0
	  mirror-0   ONLINE       0     0     0
	    nvme2n1  ONLINE       0     0     0
	    nvme3n1  ONLINE       0     0     0
	logs	
	  nvme5n1    ONLINE       0     0     0
	cache
	  nvme0n1    REMOVED      0     0     0
	spares
	  nvme1n1    AVAIL   

errors: No known data errors

# file /dev/nvme1n1
/dev/nvme1n1: cannot open `/dev/nvme1n1' (No such file or directory)
```

I also verified that in both removal cases (the cache and spare), that udev remove events were generated.  So ZED should be able to key off those removal events for spares as well, although it might require some additional code.

**Comment by ixhamza** [comment]
@tonyhutter you are right, hotplugging of spare is indeed not supported in this patch since it does not contain a label that we use to identify other vdevs.
@freqlabs - do you think this is something we may cover in future work as it does not seem to be very straightforward to cover in this PR with reference to our earlier discussion?

**Comment by freqlabs** [comment]
Yes, we can address handling spares next.

**Comment by ixhamza** [comment]
@behlendorf, @tonyhutter It would be appreciated if you guys can provide feedback or tag relevant guys here to review this PR. These patches are very critical for our TrueNAS BETA release.
@amotin, @freqlabs

**Comment by tonyhutter** [comment]
@ixhamza apologies for the delay, I'll take a look right now

File: lib/libzfs/libzfs_pool.c [hunk 9]
```diff
@@ -3073,6 +3073,39 @@ zpool_vdev_offline(zpool_handle_t *zhp, const char *path, boolean_t istmp)
 	}
 }
 
+/*
+ * Remove the specified vdev. Called from zed on udev remove event.
+ */
+int
+zpool_vdev_remove_wanted(zpool_handle_t *zhp, const char *path)
```
Comment (tonyhutter) [inline lib/libzfs/libzfs_pool.c:3080]: Both `zpool_vdev_remove_wanted()` and `zpool_vdev_remove()` are exported by libzfs and the same signature:
```C
/*
 * Remove the given device.
 */
int
zpool_vdev_remove(zpool_handle_t *zhp, const char *path)
```
Can you add some additional comments to differentiate what `zpool_vdev_remove_wanted()` does vs `zpool_vdev_remove()`?

Comment (ixhamza) [reply lib/libzfs/libzfs_pool.c:3080]: That's a very good point. ```zpool_vdev_remove_wanted()``` asynchronously removes the vdev whereas ```zpool_vdev_remove()``` does it synchronously. I will test it again to recall what issues I faced when using ```zpool_vdev_remove()``` and will add the comment accordingly.

Comment (tonyhutter) [reply lib/libzfs/libzfs_pool.c:3080]: :+1:

**Comment by tonyhutter** [comment]
@ixhamza it looks good - I don't really don't have any more comments.  I would recommend you squash your commits and use the (slightly modified) commit message: 
```
zed: mark disks as REMOVED when they are removed

ZED does not take any action for disk removal events if there is no
spare VDEV available. Added zpool_vdev_remove_wanted() in libzfs
and vdev_remove_wanted() in vdev.c to remove the VDEV through ZED
on removal event.  This means that if you are running zed and
remove a disk, it will be propertly marked as REMOVED.
```

File: lib/libzfs/libzfs_pool.c (hunk 9 above)
Comment (ixhamza) [reply lib/libzfs/libzfs_pool.c]: Just update the comment. Thanks!

**Comment by ixhamza** [comment]
@tonyhutter Thank you so much for your review and for verifying it on your hardware. I have incorporated your feedback into this PR.

**Review by tonyhutter — APPROVED** [review]

**Comment by freqlabs** [comment]
I've offered a few more suggestions in private for Ameer to incorporate, including fixing a nearby memory leak when "remove" events are ignored.

**Comment by ixhamza** [comment]
Thank you @freqlabs. I have incorporated your suggested changes.

**Review by freqlabs — APPROVED** [review]
